File: src/buf.rs
//! Add length to the actual data to judge the integrity of the data

use crate::val::ByteQue;

#[derive(Debug)]
pub struct RecvBuf {
    buf: Vec<u8>,
    size: Option<usize>,
}

impl RecvBuf {
    #[inline]
    pub fn new() -> Self {
        RecvBuf {
            buf: Vec::new(),
            size: None,
        }
    }

    pub fn append(&mut self, other: &[u8]) {
        match self.size {
            Some(s) => {
                if s > self.buf.len() {
                    let l = s - self.buf.len();
                    if l < other.len() {
                        self.buf.extend_from_slice(&other[..l]);
                    } else {
                        self.buf.extend_from_slice(other);
                    }
                }
            }
            None => {
                if self.buf.is_empty() {
                    for x in 0..other.len() {
                        // 64-bit computer will overflow if greater than 9
                        if x == 9 || other[x] <= 0x7f {
                            let mut s = 0usize;
                            for i in 0..=x {
                                s |= (other[i] as usize & 0x7f) << 7 * i;
                            }
                            self.size = Some(s);
                            let t = &other[x + 1..];
                            if s < t.len() {
                                self.buf.extend_from_slice(&t[..s]);
                            } else {
                                self.buf.extend_from_slice(t);
                            }
                            return;
                        }
                    }
                    self.buf.extend_from_slice(other);
                } else {
                    self.buf.extend_from_slice(other);
                    for x in 0..self.buf.len() {
                        // 64-bit computer will overflow if greater than 9
                        if x == 9 || self.buf[x] <= 0x7f {
                            let mut s = 0usize;
                            for i in 0..=x {
                                s |= (self.buf.remove(0) as usize & 0x7f) << 7 * i;
                            }
                            self.size = Some(s);
                            if self.buf.len() > s {
                                self.buf.resize(s, 0);
                            }
                            break;
                        }
                    }
                }
            }
        }
    }

    #[inline]
    pub fn size(&self) -> Option<usize> {
        self.size
    }

    #[inline]
    pub fn len(&self) -> usize {
        self.buf.len()
    }
}

impl From<RecvBuf> for ByteQue {
    #[inline]
    fn from(other: RecvBuf) -> Self {
        ByteQue::from(other.buf)
    }
}

pub fn send_data(q: ByteQue) -> Vec<u8> {
    let mut v = Vec::new();
    let mut s = q.len();
    // maximum number of 64-bit computers
    for _ in 0..10 {
        if s <= 0x7f {
            v.push(s as u8 & 0x7f);
            break;
        } else {
            v.push(s as u8 & 0x7f | 0x80);
        }
        s >>= 7;
    }
    v.append(&mut Vec::<u8>::from(q));
    v
}

File: lrpc-go/buf/buf.go
package buf

import "github.com/lipogem/lrpc/lrpc-go/val"

//RecvBuf actual length data
type RecvBuf struct {
	buff []byte
	size *int
}

//NewRecvBuf create a new recvbuf
func NewRecvBuf() *RecvBuf {
	return &RecvBuf{
		buff: nil,
		size: nil,
	}
}

//Append add some data to the buffer until it reaches the specified length
func (r *RecvBuf) Append(other []byte) {
	if r.size == nil {
		if len(r.buff) == 0 {
			for x := 0; x < len(other); x++ {
				if x == 9 || other[x] <= 0x7f {
					s := 0
					for i := 0; i <= x; i++ {
						s |= (int(other[i]) & 0x7f) << (7 * i)
					}
					r.size = &s
					t := other[x+1:]
					if s < len(t) {
						r.buff = append(r.buff, t[:s]...)
					} else {
						r.buff = append(r.buff, t...)
					}
					return
				}
			}
			r.buff = append(r.buff, other...)
		} else {
			r.buff = append(r.buff, other...)
			for x := 0; x < len(r.buff); x++ {
				if x == 9 || r.buff[x] <= 0x7f {
					s := 0
					for i := 0; i <= x; i++ {
						s |= (int(r.buff[i]) & 0x7f) << (7 * i)
					}
					r.buff = r.buff[x+1:]
					r.size = &s
					if len(r.buff) > s {
						r.buff = r.buff[:s]
					}
					break
				}
			}
		}
	} else {
		if *r.size > len(r.buff) {
			l := *r.size - len(r.buff)
			if l < len(other) {
				r.buff = append(r.buff, other[:l]...)
			} else {
				r.buff = append(r.buff, other...)
			}
		}
	}
}

//Size the length of data that should be received
func (r *RecvBuf) Size() *int {
	return r.size
}

//Len length of data received
func (r *RecvBuf) Len() int {
	return len(r.buff)
}

//ByteQue get ByteQue
func (r *RecvBuf) ByteQue() *val.ByteQue {
	return val.Copy(r.buff)
}

//SendData add length to the actual data to judge the integrity of the data
func SendData(q *val.ByteQue) []byte {
	n := q.Len()
	v := make([]byte, n+10)
	i := 0
	for s := n; i < 10; i++ {
		if s <= 0x7f {
			v[i] = byte(s) & 0x7f
			break
		} else {
			v[i] = byte(s)&0x7f | 0x80
		}
		s >>= 7
	}
	q.CopyTo(v[i+1:])
	return v[:i+1+n]
}
